Assets/HoloToolkit-Preview/QRTracker/Scripts/QRCodesVisualizer.cs

```diff
@@ -0,0 +1,109 @@
+﻿using System.Collections;
+
+using System.Collections.Generic;
+
+using UnityEngine;
+
+using QRCodesTrackerPlugin;
+namespace HoloToolkit.Unity.QRTracking
+{
+    public class QRCodesVisualizer : MonoBehaviour
+    {
+        public GameObject qrCodePrefab;
+
+        private System.Collections.Generic.SortedDictionary<System.Guid, GameObject> qrCodesObjectsList;
+
+        struct ActionData
+        {
+            public enum Type
+            {
+                Added,
+                Updated,
+                Removed
+            };
+            public Type type;
+            public QRCodesTrackerPlugin.QRCode qrCode;
+
+            public ActionData(Type type, QRCodesTrackerPlugin.QRCode qRCode) : this()
+            {
+                this.type = type;
+                qrCode = qRCode;
+            }
+        }
+
+        private System.Collections.Generic.Queue<ActionData> pendingActions = new Queue<ActionData>();
+        void Awake()
+        {
+            qrCodesObjectsList = new SortedDictionary<System.Guid, GameObject>();
+            QRCodesManager.Instance.QRCodeAdded += Instance_QRCodeAdded;
+            QRCodesManager.Instance.QRCodeUpdated += Instance_QRCodeUpdated;
+            QRCodesManager.Instance.QRCodeRemoved += Instance_QRCodeRemoved;
+        }
+
+        // Use this for initialization
+        void Start()
+        {
+            if (qrCodePrefab == null)
+            {
+                throw new System.Exception("Prefab not assigned");
+            }
+        }
+
+        private void Instance_QRCodeAdded(object sender, QRCodeEventArgs<QRCodesTrackerPlugin.QRCode> e)
+        {
+            lock (pendingActions)
+            {
+                pendingActions.Enqueue(new ActionData(ActionData.Type.Added, e.Data));
+            }
+        }
+
+        private void Instance_QRCodeUpdated(object sender, QRCodeEventArgs<QRCodesTrackerPlugin.QRCode> e)
+        {
+            lock (pendingActions)
+            {
+                pendingActions.Enqueue(new ActionData(ActionData.Type.Updated, e.Data));
+            }
+        }
+
+        private void Instance_QRCodeRemoved(object sender, QRCodeEventArgs<QRCodesTrackerPlugin.QRCode> e)
+        {
+            lock (pendingActions)
+            {
+                pendingActions.Enqueue(new ActionData(ActionData.Type.Removed, e.Data));
+            }
+        }
+
+        private void HandleEvents()
+        {
+            lock (pendingActions)
+            {
+                while (pendingActions.Count > 0)
+                {
+                    var action = pendingActions.Dequeue();
+                    if (action.type == ActionData.Type.Added)
```

Comment:
nit: With enums, I think switch works better. It makes it easier to understand the flow for us mere mortals